pkg/devfile/adapters/docker/component/adapter.go

```diff
@@ -0,0 +1,219 @@
+			// Container doesn't exist, so need to pull its image (to be safe) and start a new container
+			s := log.Spinner("Pulling image " + *comp.Image)
+			defer s.End(false)
+
+			err = a.Client.PullImage(*comp.Image)
+			if err != nil {
+				return errors.Wrapf(err, "Unable to pull %s image", *comp.Image)
+			}
+			s.End(true)
+
+			// Start the container
+			err = a.startContainer(componentName, projectVolumeName, comp)
+			if err != nil {
+				return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+			}
+			glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+		} else {
+			// Container already exists
+			containerID := containers[0].ID
+
+			// Get the associated container config from the container ID
+			containerConfig, err := a.Client.GetContainerConfig(containerID)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get the container config for component %s", componentName)
+			}
+
+			// See if the container needs to be updated
+			if utils.DoesContainerNeedUpdating(comp, containerConfig) {
+				glog.V(3).Info("Updating the component")
+				// Remove the container
+				err := a.Client.RemoveContainer(containerID)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to remove container %s for component %s", containerID, *comp.Alias)
+				}
+
+				// Start the container
+				err = a.startContainer(componentName, projectVolumeName, comp)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+				}
+				glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+			}
+		}
+	}
+	return nil
+}
+
+func (a Adapter) startContainer(componentName string, projectVolumeName string, comp versionsCommon.DevfileComponent) error {
+	containerConfig := a.getContainerConfig(componentName, comp)
+	hostConfig := container.HostConfig{}
+
+	// If the component set `mountSources` to true, add the source volume to it
+	if comp.MountSources {
+		addProjectVolumeToComp(projectVolumeName, &hostConfig)
+	}
+
+	// Create the docker container
+	s := log.Spinner("Starting container for " + *comp.Image)
+	defer s.End(false)
+	err := a.Client.StartContainer(&containerConfig, &hostConfig, nil)
+	if err != nil {
+		return err
+	}
+	s.End(true)
+	return nil
+}
+func (a Adapter) getContainerConfig(componentName string, comp versionsCommon.DevfileComponent) container.Config {
+	// Convert the env vars in the Devfile to the format expected by Docker
+	envVars := utils.ConvertEnvs(comp.Env)
+
+	containerLabels := map[string]string{
+		"component": componentName,
+		"alias":     *comp.Alias,
+	}
+
+	containerConfig := a.Client.GenerateContainerConfig(*comp.Image, comp.Command, comp.Args, envVars, containerLabels)
+	return containerConfig
+}
+
+func addProjectVolumeToComp(projectVolumeName string, hostConfig *container.HostConfig) *container.HostConfig {
```

Comment:
Can this be moved to docker/utils/utils.go to be consistent with kubernetes package?

Reply:
Makes sense to me 👍

Reply:
Updated.